source/exe/main_common.cc

```diff
@@ -21,79 +22,111 @@
+      // hot restart at the global level. The per-listener drain managers decide whether to
+      // to include /healthcheck/fail status.
+      new Server::DrainManagerImpl(server, envoy::api::v2::Listener_DrainType_MODIFY_ONLY)};
+}
 
-} // namespace Server
+Runtime::LoaderPtr ProdComponentFactory::createRuntime(Server::Instance& server,
+                                                       Server::Configuration::Initial& config) {
+  return Server::InstanceUtil::createRuntime(server, config);
+}
 
-int main_common(OptionsImpl& options) {
-  Stats::RawStatData::configure(options);
+MainCommonBase::MainCommonBase(OptionsImpl& options, bool hot_restart) : options_(options) {
+  ares_library_init(ARES_LIB_INIT_ALL);
+  Event::Libevent::Global::initialize();
+  RELEASE_ASSERT(Envoy::Server::validateProtoDescriptors());
 
+  switch (options_.mode()) {
+  case Server::Mode::Serve: {
 #ifdef ENVOY_HOT_RESTART
-  std::unique_ptr<Server::HotRestartImpl> restarter;
-  try {
-    restarter.reset(new Server::HotRestartImpl(options));
-  } catch (Envoy::EnvoyException& e) {
-    std::cerr << "unable to initialize hot restart: " << e.what() << std::endl;
-    return 1;
+    if (hot_restart) {
+      restarter_.reset(new Server::HotRestartImpl(options_));
+    }
+#endif
+    if (!hot_restart) {
+      restarter_.reset(new Server::HotRestartNopImpl());
+    }
+
+    Stats::RawStatData::configure(options_);
+    tls_.reset(new ThreadLocal::InstanceImpl);
+    Thread::BasicLockable& log_lock = restarter_->logLock();
+    Thread::BasicLockable& access_log_lock = restarter_->accessLogLock();
+    auto local_address = Network::Utility::getLocalAddress(options_.localAddressIpVersion());
+    Logger::Registry::initialize(options_.logLevel(), log_lock);
+
+    stats_store_.reset(new Stats::ThreadLocalStoreImpl(restarter_->stats_allocator()));
+    server_.reset(new Server::InstanceImpl(options_, local_address, default_test_hooks_,
+                                           *restarter_, *stats_store_, access_log_lock,
+                                           component_factory_, *tls_));
+    break;
   }
+  case Server::Mode::Validate:
+    break;
+  }
+}
 
-  Thread::BasicLockable& log_lock = restarter->logLock();
-  Thread::BasicLockable& access_log_lock = restarter->accessLogLock();
-  Stats::RawStatDataAllocator& stats_allocator = *restarter;
-#else
-  std::unique_ptr<Server::HotRestartNopImpl> restarter;
-  restarter.reset(new Server::HotRestartNopImpl());
-
-  Thread::MutexBasicLockable log_lock, access_log_lock;
-  Stats::HeapRawStatDataAllocator stats_allocator;
-#endif
+MainCommonBase::~MainCommonBase() { ares_library_cleanup(); }
 
-  RELEASE_ASSERT(Envoy::Server::validateProtoDescriptors());
-  Event::Libevent::Global::initialize();
-  Server::ProdComponentFactory component_factory;
-  auto local_address = Network::Utility::getLocalAddress(options.localAddressIpVersion());
-  switch (options.mode()) {
+bool MainCommonBase::run() {
```

Comment:
Should we have a better name for this? This capture historically what was in the common main, but I think logically it's about overall server setup, teardown and run. We already have a `Server::Instance`, so maybe `Server::Cli`, `Server::Launcher`, etc.?

Reply:
I'm OK renaming it but that would also lose any kind of a/b correlation in the review.  AFAIK there's no way to tell git that a file was renamed; it either figures it out or it doesn't, and mostly it doesn't :)

Can this be a follow-up?

Reply:
Sure. Either later or in a final pass before merging, either works.

Reply:
Thanks; I'll put this in a follow-up.